### ISSUE TITLE
javascript: WebElement.isElementPresent() promise does not resolve with `false`

### ISSUE THREAD

**Issue by L1NT** [SeleniumHQ/selenium#1641]
Example code:
>        driver.findElement(By.css('.container')).then((webElement) => {
>                webElement.isElementPresent(By.css('.foo'))
>                    .then((found) => { console.log(found); });
>            });

Assuming that an element with a class of 'container' exists, if an element with class 'foo' is present, then `found == true` and it's logged to the console. If an element with class 'foo' does not exist, then the Promise callback never executes.

I believe this had worked as expected in the past (maybe in Nov 2015 - unsure of what version I was running at the time), but was unsuccessful in identifying any code changes that would have impacted this behavior during my initial cursory search.

**Comment by jleyba** [comment]
What browser and Selenium version are you using? I'm not able to reproduce this with 2.52.0 on Firefox or Chrome:

```javascript
var webdriver = require('selenium-webdriver'),
    By = webdriver.By,
    test = require('selenium-webdriver/testing'),
    assert = require('selenium-webdriver/testing/assert');

describe('isElementPresent', function() {
  var driver;

  test.beforeEach(function() {
    driver = new webdriver.Builder()
        .forBrowser('firefox')
        .build();
  });

  test.afterEach(function() {
    driver.quit();
  });

  test.it('', function() {
    driver.get('data:text/html,%3Cdiv%3E%3Cdiv%20class%3D%22foo%22%3Ehi%3C%2Fdiv%3E%3C%2Fdiv%3E');

    var el = driver.findElement(By.css('div'));
    assert(el.isElementPresent(By.css('.bar'))).isFalse();
    assert(el.isElementPresent(By.css('.foo'))).isTrue();
    assert(el.findElement(By.css('.foo')).getText()).isEqualTo('hi');
  });
});

```

**Comment by L1NT** [comment]
I think this may be worthy of closing under the premise that "L1NT has bigger issues".

In particular, I was creating a complete generic test case to provide as an example to demonstrate that `isElementPresent()` resolves for true conditions but does not when false, however I cannot get my `true` example to work as expected either (but it still does within my actual testing code). This leads me to believe I'm doing something much more fundamentally wrong.

Here are the details of my setup:
> selenium-server-standalone-2.48.2.jar

npm list --depth=0 (edited to pertinent packages)
>├── jasmine@2.3.2
>├── jasmine-core@2.3.4
>├── selenium-webdriver@2.48.2

node --version
> v5.6.0

Example that times out at the `isElementPresent()` promise handler for both true _and_ false cases (unlike my original report):

```
// execute with (jasmine.json contains little more than the spec directory and file patterns):
// jasmine --verbose --stop-on-failure=false JASMINE_CONFIG_PATH=t/js/selenium/jasmine.json
/**
 * selenium.html:
 * <html>
 * <head></head>
 * <body>
 *     <div class='container'>
 *         <div class='foo'>foo</div>
 *     </div>
 * </body>
 * </html>
 */
describe('is element present', function() {
    var webdriver = require('selenium-webdriver');
    var By = webdriver.By;
    var until = webdriver.until;
    var driver = new webdriver.Builder()
        .withCapabilities(webdriver.Capabilities.chrome())
       .build();

    beforeEach(function(done) {
        driver.get('http://localhost/selenium.html');
    });
    afterAll(function(done) {
        driver.quit().then(done);
    });

    it('should find .foo', function(done) {
        driver.wait(until.elementLocated(By.css('.container')))
            .then((el) => {
                el.isElementPresent(By.css('.foo'))
                    .then(function(found) {
                        expect(found).toBe(true);
                        done();
                    });
            });
    });
    it('should not find .bar', function(done) {
        driver.wait(until.elementLocated(By.css('.container')))
            .then(function(el) {
                el.isElementPresent(By.css('.bar'))
                    .then((found) => {
                        expect(found).toBe(false);
                        done();
                    });
            });
    });
});
```
Reviewing my setup, I wonder if I'm experiencing a problem introduced by my Nodejs version or using Jasmine (a name-collision over the `describe()` function, maybe?). Thanks for taking the time to look at my report! I'll be sure to update if I can uncover what I'm doing wrong.

**Comment by jleyba** [comment]
If your `expect(...)` statement fails, the exception will be caught by the promise and `done()` will never be called.

```javascript
el.isElementPresent(By.css('.bar')).then(found => {
  expect(found).toBe(false);
  done();
});
```

Should be something like:

```javascript
el.isElementPresent(By.css('.bar'))
    .then(found => expect(found).toBe(false))
    .then(done, done);
```

With mocha you can return a promise from the test function instead of having to call `done()`. Is that possible with Jasmine?

I'm closing this because the Selenium API is functioning properly. If you need further help, you should try the selenium users mailing list or #selenium IRC channel.

As a final note, be aware that `isElementPresent` is now deprecated (3dfc66be727369f2d9083a7e69bb2f2b12d1c440) and will be removed in Selenium 3.0

**Comment by L1NT** [comment]
@jleyba Thanks again for taking the time to look into my report. Because of your investment in helping me, I feel compelled to share my error now that I've discovered what I was doing wrong:

Jasmine executes its specs asynchronously, and if passed a callback function (`done`) will abort if a pre-determined timeout is reached before `done()` is called. Basically, as I added more tasks to perform within the spec and more specs, my single-connection dev server couldn't process the requests within the time allotted by the Jasmine timeout. As suggested, I need to restructure my code to rely on promises instead of callbacks that are time-constrained.

Cheers!
